"process didn't exit successfully" output should use the same quoting as verbose command output

When writing rustc arguments like `--cfg feature="default"`, the doublequotes matter to rustc, so cargo needs to write that out as `--cfg 'feature="default"'`; otherwise, pasting the command line into a shell will eat the doublequotes and rustc won't accept the result.

Related to https://github.com/rust-lang/cargo/pull/1770 ; looks like this didn't get covered by that fix.

EDIT: based on the further input from @mikeyhew below, the problem isn't that `cargo build -v` output is going unquoted, it's that the "process didn't exit successfully" message needs the same treatment.

OK I managed to reproduce this. With `cargo build/check -v`, the shell command that it prints out *before* running is quoted properly. It's the command that is printed out after compile fails that isn't quoted, and I that's the one I copied and pasted, and kept looking at for the crate that originally had the problem.

Here is an example output:

```
Michaels-MacBook-Air-2:foo* mikeyhew$ cargo check -v
    Checking foo v0.1.0 (file:///Users/mikeyhew/Desktop/Projects/foo)
     Running `rustc --crate-name foo src/main.rs --crate-type bin --emit=dep-info,metadata -C debuginfo=2 --cfg 'feature="default"' --cfg 'feature="some_feature"' -C metadata=3af84aca968e8c04 -C extra-filename=-3af84aca968e8c04 --out-dir /Users/mikeyhew/Desktop/Projects/foo/target/debug/deps -C incremental=/Users/mikeyhew/Desktop/Projects/foo/target/debug/incremental -L dependency=/Users/mikeyhew/Desktop/Projects/foo/target/debug/deps`
error[E0425]: cannot find value `rust` in this scope
 --> src/main.rs:2:11
  |
2 |     break rust;
  |           ^^^^ not found in this scope

error[E0268]: `break` outside of loop
 --> src/main.rs:2:5
  |
2 |     break rust;
  |     ^^^^^^^^^^ cannot break outside of a loop

error: internal compiler error: It looks like you're trying to break rust; would you like some ICE?

note: the compiler expectedly panicked. this is a feature.

note: we would appreciate a joke overview: https://github.com/rust-lang/rust/issues/43162#issuecomment-320764675

note: rustc 1.28.0-nightly (5205ae8bd 2018-06-13) running on x86_64-apple-darwin

error: aborting due to 3 previous errors

Some errors occurred: E0268, E0425.
For more information about an error, try `rustc --explain E0268`.
error: Could not compile `foo`.

Caused by:
  process didn't exit successfully: `rustc --crate-name foo src/main.rs --crate-type bin --emit=dep-info,metadata -C debuginfo=2 --cfg feature="default" --cfg feature="some_feature" -C metadata=3af84aca968e8c04 -C extra-filename=-3af84aca968e8c04 --out-dir /Users/mikeyhew/Desktop/Projects/foo/target/debug/deps -C incremental=/Users/mikeyhew/Desktop/Projects/foo/target/debug/incremental -L dependency=/Users/mikeyhew/Desktop/Projects/foo/target/debug/deps` (exit code: 101)
```

Take a look at the command printed out before attempting to compile/check

```
Checking foo v0.1.0 (file:///Users/mikeyhew/Desktop/Projects/foo)
     Running `rustc --crate-name foo src/main.rs --crate-type bin --emit=dep-info,metadata -C debuginfo=2 --cfg 'feature="default"' --cfg 'feature="some_feature"' -C metadata=3af84aca968e8c04 -C extra-filename=-3af84aca968e8c04 --out-dir /Users/mikeyhew/Desktop/Projects/foo/target/debug/deps -C incremental=/Users/mikeyhew/Desktop/Projects/foo/target/debug/incremental -L dependency=/Users/mikeyhew/Desktop/Projects/foo/target/debug/deps`
```

and the command printed out after it fails:

```
  process didn't exit successfully: `rustc --crate-name foo src/main.rs --crate-type bin --emit=dep-info,metadata -C debuginfo=2 --cfg feature="default" --cfg feature="some_feature" -C metadata=3af84aca968e8c04 -C extra-filename=-3af84aca968e8c04 --out-dir /Users/mikeyhew/Desktop/Projects/foo/target/debug/deps -C incremental=/Users/mikeyhew/Desktop/Projects/foo/target/debug/incremental -L dependency=/Users/mikeyhew/Desktop/Projects/foo/target/debug/deps` (exit code: 101)
```

In the failure message, the `--cfg feature="default"` option is not quoted properly.

--- a/src/cargo/util/process_builder.rs
+++ b/src/cargo/util/process_builder.rs
@@ -133,7 +133,7 @@ impl ProcessBuilder {
         let mut command = self.build_command();
         let exit = command.status().chain_err(|| {
             process_error(
-                &format!("could not execute process `{}`", self.debug_string()),
+                &format!("could not execute process {}", self),
                 None,
                 None,
             )
@@ -143,10 +143,7 @@ impl ProcessBuilder {
             Ok(())
         } else {
             Err(process_error(
-                &format!(
-                    "process didn't exit successfully: `{}`",
-                    self.debug_string()
-                ),
+                &format!("process didn't exit successfully: {}", self),
                 Some(&exit),
                 None,
             ).into())
@@ -164,7 +161,7 @@ impl ProcessBuilder {
         let error = command.exec();
         Err(CargoError::from(error)
             .context(process_error(
-                &format!("could not execute process `{}`", self.debug_string()),
+                &format!("could not execute process {}", self),
                 None,
                 None,
             ))
@@ -185,7 +182,7 @@ impl ProcessBuilder {
 
         let output = command.output().chain_err(|| {
             process_error(
-                &format!("could not execute process `{}`", self.debug_string()),
+                &format!("could not execute process {}", self),
                 None,
                 None,
             )
@@ -195,10 +192,7 @@ impl ProcessBuilder {
             Ok(output)
         } else {
             Err(process_error(
-                &format!(
-                    "process didn't exit successfully: `{}`",
-                    self.debug_string()
-                ),
+                &format!("process didn't exit successfully: {}", self),
                 Some(&output.status),
                 Some(&output),
             ).into())
@@ -261,7 +255,7 @@ impl ProcessBuilder {
         })()
             .chain_err(|| {
             process_error(
-                &format!("could not execute process `{}`", self.debug_string()),
+                &format!("could not execute process {}", self),
                 None,
                 None,
             )
@@ -276,16 +270,13 @@ impl ProcessBuilder {
             let to_print = if print_output { Some(&output) } else { None };
             if !output.status.success() {
                 return Err(process_error(
-                    &format!(
-                        "process didn't exit successfully: `{}`",
-                        self.debug_string()
-                    ),
+                    &format!("process didn't exit successfully: {}", self),
                     Some(&output.status),
                     to_print,
                 ).into());
             } else if let Some(e) = callback_error {
                 let cx = process_error(
-                    &format!("failed to parse process output: `{}`", self.debug_string()),
+                    &format!("failed to parse process output: {}", self),
                     Some(&output.status),
                     to_print,
                 );
@@ -320,16 +311,6 @@ impl ProcessBuilder {
             c.configure(&mut command);
         }
         command
-    }
-
-    /// Get the command line for the process as a string.
-    fn debug_string(&self) -> String {
-        let mut program = format!("{}", self.program.to_string_lossy());
-        for arg in &self.args {
-            program.push(' ');
-            program.push_str(&format!("{}", arg.to_string_lossy()));
-        }
-        program
     }
 }
 

--- a/tests/testsuite/shell_quoting.rs
+++ b/tests/testsuite/shell_quoting.rs
@@ -1,0 +1,43 @@
+//! this file tests that when the commands being run are shown
+//! in the output, their arguments are quoted properly
+//! so that the command can be run in a terminal
+
+use cargotest::support::{
+    execs,
+    project,
+};
+use hamcrest::assert_that
+
+#[test]
+fn features_are_quoted() {
+    let p = project("foo")
+        .file(
+            "Cargo.toml",
+            r#"
+            [project]
+            name = "foo"
+            version = "0.1.0"
+            authors = ["mikeyhew@example.com"]
+
+            [features]
+            some_feature = []
+            default = ["some_feature"]
+            "#,
+        )
+        .file("src/main.rs", "fn main() {error}")
+        .build();
+
+    assert_that(
+        p.cargo("check -v"),
+        execs()
+            .with_status(101)
+            .with_stderr_contains(
+                r#"\
+[CHECKING] foo [..]
+[RUNNING] `rustc [..] --cfg 'feature="default"' --cfg 'feature="some_feature"' [..]`
+[ERROR] [..]
+process didn't exit successfully: `rustc [..] --cfg 'feature="default"' --cfg 'feature="some_feature"' [..]`
+"#
+            )
+    );
+}

